### PR TITLE
[Rapid7 InsightIDR] Fixed an issue with receiving the logs by query

### PR DESCRIPTION
**Pull request by mmhw**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
Fixes: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-6152)

## Description
Fixed an issue with receiving the logs by query with the following commands:
* **rapid7-insight-idr-query-log**
* **rapid7-insight-idr-query-log-set**

### CODE REVIEW

File: Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR_test.py [hunk 1]
```diff
@@ -459,14 +462,113 @@ def test_insight_idr_query_log_set(requests_mock) -> None:
...
+            'Authentication': 'apikey'
+        },
+        proxy=False
+    )
+    response = commands[end_point](client, 'x', '')
+
+    outputs = []
+
+    for result in response.raw_response:
+        outputs.extend(iter(result.get('events', [])))
```
Comment (kobymeir) [inline Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR_test.py:566]: can be simplified to a list-comprehension

File: Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR_test.py [hunk 2]
```diff
@@ -421,8 +424,8 @@ def test_insight_idr_query_log(requests_mock) -> None:
     response = insight_idr_query_log_command(client, 'x', '', '', '')
 
     outputs = []
-    for event in response.raw_response.get('events', []):
-        outputs.append(event)
+    for result in response.raw_response:
+        outputs.extend(iter(result.get('events', [])))
```
Comment (kobymeir) [inline Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR_test.py:428]: also simplify as a list-comprehension

File: Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR.py [hunk 3]
```diff
@@ -667,6 +631,31 @@ def insight_idr_query_log_set_command(client: Client, log_set_id: str, query: st
...
+    results_list = [result]
+    while results_list:
+        results = results_list.pop(0)
+
+        if events := results.get('events', []):
+            data_for_readable_output.extend(events)
+            raw_responcse.append(results)
+
+        if 'links' in results:
+            for link in results.get('links', []):
```
Comment (kobymeir) [inline Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR.py:647]:
```suggestion
        if links := results.get('links', []):
            for link in links:
```

File: Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR.py [hunk 4]
```diff
@@ -667,6 +631,31 @@ def insight_idr_query_log_set_command(client: Client, log_set_id: str, query: st
...
+                url = link.get('href')
+                new_results = client.query_log_callback(url)
+                results_list.append(new_results)
+
+                events_len = len(results.get('events', []))
+                progress = results.get('progress')
+                demisto.debug(f'Events length: {events_len}, progress: {progress}')
```
Comment (kobymeir) [inline Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR.py:654]:
```suggestion
                progress = results.get('progress')
                demisto.debug(f'Events length: {len(events)}, progress: {progress}')
```
events is already assigned a value in the previous if statement

**Review by kobymeir — CHANGES_REQUESTED** [review]

File: Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/test_data/query_log_set_page_1.json [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        "links": [
+          {
+            "rel": "Context",
+            "href": "https://us.api.insight.rapid7.com/log_search/query/context/123?per_page=50×tamp=1604849794718&log_keys=7efaf894-cf8a-4ed2-9495-77395bf2e5a6&context_type=SURROUND"
+          }
+        ],
+        "sequence_number_str": "123"
+      }
+    ],
+    "links": [
```
[the remark this answers is not in the document]
Comment (mmhw) [reply Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/test_data/query_log_set_page_1.json:30]: @kobymeir I check it in the second function (test_insight_idr_query_log_with_callback), do you think it should be added here as well?

**Review by kobymeir — APPROVED** [review]

File: Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR.py [hunk 6]
```diff
@@ -667,6 +631,30 @@ def insight_idr_query_log_set_command(client: Client, log_set_id: str, query: st
     return command_results
 
 
+def handle_query_log_results(client: Client, result):
```
Comment (bziser) [inline Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR.py:634]: please add description to the function

**Review by bziser — APPROVED** [review]
great, just add the description to the function

File: Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR.py [hunk 7]
```diff
@@ -667,6 +631,41 @@ def insight_idr_query_log_set_command(client: Client, log_set_id: str, query: st
     return command_results
 
 
+def handle_query_log_results(client: Client, result: dict) -> Tuple[list, list]:
+    """This function get the first result of the query, then handles if the query has in progress, and handle pagination.
```
Comment (bziser) [inline Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR.py:635]:
```suggestion
    """This function get the first result of the query, then handles if the query is still in progress, and handle pagination.
```

File: Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR.py [hunk 8]
```diff
@@ -667,6 +631,41 @@ def insight_idr_query_log_set_command(client: Client, log_set_id: str, query: st
...
+
+    Args:
+        client (Client): Rapid7 Client
+        result (dict): The first result of the query
+
+    Returns:
+        Tuple[list, list]:
+            data_for_readable_output: The data for the readable output was contains all the events
```
Comment (bziser) [inline Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR.py:643]:
```suggestion
            data_for_readable_output: The data for the readable output which contains all the events
```

File: Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR.py [hunk 9]
```diff
@@ -667,6 +631,41 @@ def insight_idr_query_log_set_command(client: Client, log_set_id: str, query: st
...
+            raw_responcse: The raw response from the all the requests that returned events
```
Comment (bziser) [inline Packs/Rapid7_InsightIDR/Integrations/Rapid7_InsightIDR/Rapid7_InsightIDR.py:644]:
```suggestion
            raw_responcse: Raw response from all events returned by the requests
```

**Pull request merged**